Assets/HoloToolkit/Input/Scripts/Voice/KeywordManager.cs

```diff
@@ -44,23 +44,36 @@ public enum RecognizerStartBehavior { AutoStart, ManualStart };
         public KeywordAndResponse[] KeywordsAndResponses;
 
         private KeywordRecognizer keywordRecognizer;
-        private Dictionary<string, UnityEvent> responses;
+        private readonly Dictionary<string, UnityEvent> responses = new Dictionary<string, UnityEvent>();
 
         void Start()
         {
-            if (KeywordsAndResponses.Length > 0)
+            int keywordCount = KeywordsAndResponses.Length;
+            if (keywordCount > 0)
             {
-                // Convert the struct array into a dictionary, with the keywords and the keys and the methods as the values.
-                // This helps easily link the keyword recognized to the UnityEvent to be invoked.
-                responses = KeywordsAndResponses.ToDictionary(keywordAndResponse => keywordAndResponse.Keyword,
-                                                              keywordAndResponse => keywordAndResponse.Response);
-
-                keywordRecognizer = new KeywordRecognizer(responses.Keys.ToArray());
-                keywordRecognizer.OnPhraseRecognized += KeywordRecognizer_OnPhraseRecognized;
-
-                if (RecognizerStart == RecognizerStartBehavior.AutoStart)
+                try
                 {
-                    keywordRecognizer.Start();
+                    string[] keywords = new string[keywordCount];
+                    // Convert the struct array into a dictionary, with the keywords and the keys and the methods as the values.
+                    // This helps easily link the keyword recognized to the UnityEvent to be invoked.
+                    for (int index = keywordCount; --index >= 0;)
```

Comment:
I believe Linq can be "naughty" in Update() but not in Start(). I also believe that var, Linq and Reactive Extensions are more readable, once you get used to. I'm just following the project's rules here. ;)
In this case, there's a small performance gain as the array is iterated only once to fill both the dictionary and the keywords array.
Avoidance of 'foreach' has also been one of the common suggestions so I decided to try the reverse order for loop with prefixed decrement, that I usually use in C++ but never in C#. Foreach loop is [very performant on arrays](http://stackoverflow.com/questions/365615/in-net-which-loop-runs-faster-for-or-foreach/472258#472258).